<<<<<<< HEAD
# Enforce or disallow the use of the record type (`consistent-indexed-object-style`)

TypeScript supports defining object show keys can be flexible using an index signature. TypeScript also has a builtin type named `Record` to create an empty object defining only an index signature. For example, the following types are equal:
=======
---
description: 'Require or disallow the `Record` type.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/consistent-indexed-object-style** for documentation.

TypeScript supports defining arbitrary object keys using an index signature. TypeScript also has a builtin type named `Record` to create an empty object defining only an index signature. For example, the following types are equal:
>>>>>>> features/task-display-and-function

```ts
interface Foo {
  [key: string]: unknown;
}

type Foo = {
  [key: string]: unknown;
};

type Foo = Record<string, unknown>;
```

<<<<<<< HEAD
## Options

- `"record"`: Set to `"record"` to only allow the `Record` type. Set to `"index-signature"` to only allow index signatures. (Defaults to `"record"`)

For example:

```CJSON
{
    "@typescript-eslint/consistent-indexed-object-style": ["error", "index-signature"]
}
```

## Rule details

This rule enforces a consistent way to define records.

Examples of **incorrect** code with `record` option.

```ts
=======
Keeping to one declaration form consistently improve code readability.

## Options

- `"record"` _(default)_: only allow the `Record` type.
- `"index-signature"`: only allow index signatures.

### `record`

<!--tabs-->

#### ‚ùå Incorrect

```ts
/* eslint @typescript-eslint/consistent-indexed-object-style: ["error", "record"] */

>>>>>>> features/task-display-and-function
interface Foo {
  [key: string]: unknown;
}

type Foo = {
  [key: string]: unknown;
};
```

<<<<<<< HEAD
Examples of **correct** code with `record` option.

```ts
type Foo = Record<string, unknown>;
```

Examples of **incorrect** code with `index-signature` option.

```ts
type Foo = Record<string, unknown>;
```

Examples of **correct** code with `index-signature` option.

```ts
=======
#### ‚úÖ Correct

```ts
/* eslint @typescript-eslint/consistent-indexed-object-style: ["error", "record"] */

type Foo = Record<string, unknown>;
```

### `index-signature`

<!--tabs-->

#### ‚ùå Incorrect

```ts
/* eslint @typescript-eslint/consistent-indexed-object-style: ["error", "index-signature"] */

type Foo = Record<string, unknown>;
```

#### ‚úÖ Correct

```ts
/* eslint @typescript-eslint/consistent-indexed-object-style: ["error", "index-signature"] */

>>>>>>> features/task-display-and-function
interface Foo {
  [key: string]: unknown;
}

type Foo = {
  [key: string]: unknown;
};
```
