<<<<<<< HEAD
# Consistent with type definition either `interface` or `type` (`consistent-type-definitions`)

There are two ways to define a type.
=======
---
description: 'Enforce type definitions to consistently use either `interface` or `type`.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/consistent-type-definitions** for documentation.

TypeScript provides two common ways to define an object type: `interface` and `type`.
>>>>>>> features/task-display-and-function

```ts
// type alias
type T1 = {
  a: string;
  b: number;
};

// interface keyword
interface T2 {
  a: string;
  b: number;
}
```

<<<<<<< HEAD
## Options

This rule accepts one string option:

- `"interface"`: enforce using `interface`s for object type definitions.
- `"type"`: enforce using `type`s for object type definitions.

For example:

```jsonc
{
  // Use type for object definitions
  "@typescript-eslint/consistent-type-definitions": ["error", "type"]
}
```

## Rule Details

Examples of **incorrect** code with `interface` option.

```ts
type T = { x: number };
```

Examples of **correct** code with `interface` option.

```ts
=======
The two are generally very similar, and can often be used interchangeably.
Using the same type declaration style consistently helps with code readability.

## Options

- `"interface"` _(default)_: enforce using `interface`s for object type definitions.
- `"type"`: enforce using `type`s for object type definitions.

### `interface`

<!--tabs-->

#### ❌ Incorrect

```ts
/* eslint @typescript-eslint/consistent-type-definitions: ["error", "interface"] */

type T = { x: number };
```

#### ✅ Correct

```ts
/* eslint @typescript-eslint/consistent-type-definitions: ["error", "interface"] */

>>>>>>> features/task-display-and-function
type T = string;
type Foo = string | {};

interface T {
  x: number;
}
```

<<<<<<< HEAD
Examples of **incorrect** code with `type` option.

```ts
=======
### `type`

<!--tabs-->

#### ❌ Incorrect

```ts
/* eslint @typescript-eslint/consistent-type-definitions: ["error", "type"] */

>>>>>>> features/task-display-and-function
interface T {
  x: number;
}
```

<<<<<<< HEAD
Examples of **correct** code with `type` option.

```ts
=======
#### ✅ Correct

```ts
/* eslint @typescript-eslint/consistent-type-definitions: ["error", "type"] */

>>>>>>> features/task-display-and-function
type T = { x: number };
```

## When Not To Use It

If you specifically want to use an interface or type literal for stylistic reasons, you can disable this rule.
<<<<<<< HEAD

## Compatibility

- TSLint: [interface-over-type-literal](https://palantir.github.io/tslint/rules/interface-over-type-literal/)
=======
>>>>>>> features/task-display-and-function
