<<<<<<< HEAD
# Disallow non-null assertion in locations that may be confusing (`no-confusing-non-null-assertion`)

## Rule Details
=======
---
description: 'Disallow non-null assertion in locations that may be confusing.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-confusing-non-null-assertion** for documentation.
>>>>>>> features/task-display-and-function

Using a non-null assertion (`!`) next to an assign or equals check (`=` or `==` or `===`) creates code that is confusing as it looks similar to a not equals check (`!=` `!==`).

```typescript
a! == b; // a non-null assertions(`!`) and an equals test(`==`)
a !== b; // not equals test(`!==`)
a! === b; // a non-null assertions(`!`) and an triple equals test(`===`)
```

<<<<<<< HEAD
Examples of **incorrect** code for this rule:
=======
This rule flags confusing `!` assertions and suggests either removing them or wrapping the asserted expression in `()` parenthesis.

## Examples

<!--tabs-->

### âŒ Incorrect
>>>>>>> features/task-display-and-function

```ts
interface Foo {
  bar?: string;
  num?: number;
}

const foo: Foo = getFoo();
const isEqualsBar = foo.bar! == 'hello';
const isEqualsNum = 1 + foo.num! == 2;
```

<<<<<<< HEAD
Examples of **correct** code for this rule:
=======
### âœ… Correct
>>>>>>> features/task-display-and-function

<!-- prettier-ignore -->
```ts
interface Foo {
  bar?: string;
  num?: number;
}

const foo: Foo = getFoo();
const isEqualsBar = foo.bar == 'hello';
const isEqualsNum = (1 + foo.num!) == 2;
```

## When Not To Use It

If you don't care about this confusion, then you will not need this rule.

## Further Reading

- [`Issue: Easy misunderstanding: "! ==="`](https://github.com/microsoft/TypeScript/issues/37837) in [TypeScript repo](https://github.com/microsoft/TypeScript)
