<<<<<<< HEAD
# Disallow usage of the implicit `any` type in catch clauses (`no-implicit-any-catch`)
=======
---
description: 'Disallow usage of the implicit `any` type in catch clauses.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-implicit-any-catch** for documentation.

:::danger Deprecated

This rule has been deprecated as TypeScript versions >=4 includes a `useUnknownInCatchVariables` compiler option with the same check.
:::
>>>>>>> features/task-display-and-function

TypeScript 4.0 added support for adding an explicit `any` or `unknown` type annotation on a catch clause variable.

By default, TypeScript will type a catch clause variable as `any`, so explicitly annotating it as `unknown` can add a lot of safety to your codebase.

<<<<<<< HEAD
The `noImplicitAny` flag in TypeScript does not cover this for backwards compatibility reasons.

## Rule Details

This rule requires an explicit type to be declared on a catch clause variable.

The following pattern is considered a warning:
=======
The `noImplicitAny` flag in TypeScript does not cover this for backwards compatibility reasons, however you can use `useUnknownInCatchVariables` (part of `strict`) instead of this rule.

## DEPRECATED

## Examples

This rule requires an explicit type to be declared on a catch clause variable.

<!--tabs-->

### âŒ Incorrect
>>>>>>> features/task-display-and-function

```ts
try {
  // ...
} catch (e) {
  // ...
}
```

<<<<<<< HEAD
The following pattern is **_not_** considered a warning:
=======
### âœ… Correct
>>>>>>> features/task-display-and-function

<!-- TODO: prettier currently removes the type annotations, re-enable this once prettier is updated -->
<!-- prettier-ignore-start -->

```ts
try {
  // ...
} catch (e: unknown) {
  // ...
}
```

<!-- prettier-ignore-end -->

## Options

<<<<<<< HEAD
The rule accepts an options object with the following properties:

```ts
type Options = {
  // if false, disallow specifying `: any` as the error type as well. See also `no-explicit-any`
  allowExplicitAny: boolean;
};

const defaults = {
  allowExplicitAny: false,
};
```

=======
>>>>>>> features/task-display-and-function
### `allowExplicitAny`

The follow is is **_not_** considered a warning with `{ allowExplicitAny: true }`

<<<<<<< HEAD
<!-- TODO: prettier currently removes the type annotations, re-enable this once prettier is updated -->
<!-- prettier-ignore-start -->

=======
>>>>>>> features/task-display-and-function
```ts
try {
  // ...
} catch (e: any) {
  // ...
}
```

<<<<<<< HEAD
<!-- prettier-ignore-end -->

=======
>>>>>>> features/task-display-and-function
## When Not To Use It

If you are not using TypeScript 4.0 (or greater), then you will not be able to use this rule, annotations on catch clauses is not supported.

## Further Reading

<<<<<<< HEAD
- [TypeScript 4.0 Beta Release Notes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0-beta/#unknown-on-catch)
=======
- [TypeScript 4.0 Release Notes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#unknown-on-catch)
>>>>>>> features/task-display-and-function
