<<<<<<< HEAD
# require or disallow padding lines between statements (`padding-line-between-statements`)

## Rule Details

This rule extends the base [`eslint/padding-line-between-statements`](https://eslint.org/docs/rules/padding-line-between-statements) rule.

**It adds support for TypeScript constructs such as `interface` and `type`.**

## How to use

```jsonc
{
  // note you must disable the base rule as it can report incorrect errors
  "padding-line-between-statements": "off",
  "@typescript-eslint/padding-line-between-statements": [
    "error",
    {
      "blankLine": "always",
      "prev": "var",
      "next": "return"
    }
  ]
}
```
=======
---
description: 'Require or disallow padding lines between statements.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/padding-line-between-statements** for documentation.

## Examples

This rule extends the base [`eslint/padding-line-between-statements`](https://eslint.org/docs/rules/padding-line-between-statements) rule.
It adds support for TypeScript constructs such as `interface` and `type`.

## Options

In addition to options provided by ESLint, `interface` and `type` can be used as statement types.

For example, to add blank lines before interfaces and type definitions:
>>>>>>> features/task-display-and-function

```jsonc
{
  // Example - Add blank lines before interface and type definitions.
<<<<<<< HEAD
  // note you must disable the base rule as it can report incorrect errors
=======
>>>>>>> features/task-display-and-function
  "padding-line-between-statements": "off",
  "@typescript-eslint/padding-line-between-statements": [
    "error",
    {
      "blankLine": "always",
      "prev": "*",
      "next": ["interface", "type"]
    }
  ]
}
```

<<<<<<< HEAD
## Options

See [`eslint/padding-line-between-statements` options](https://eslint.org/docs/rules/padding-line-between-statements#options).

**Note** - In addition to options provided by ESLint, we have also added options for `interface` and `type`.

<sup>Taken with ‚ù§Ô∏è [from ESLint core](https://github.com/eslint/eslint/blob/master/docs/rules/padding-line-between-statements.md)</sup>
=======
**Note:** ESLint `cjs-export` and `cjs-import` statement types are renamed to `exports` and `require` respectively.
>>>>>>> features/task-display-and-function
