<<<<<<< HEAD
# Use function types instead of interfaces with call signatures (`prefer-function-type`)

## Rule Details

This rule suggests using a function type instead of an interface or object type literal with a single call signature.

Examples of **incorrect** code for this rule:

```ts
interface Foo {
=======
---
description: 'Enforce using function types instead of interfaces with call signatures.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/prefer-function-type** for documentation.

TypeScript allows for two common ways to declare a type for a function:

- Function type: `() => string`
- Object type with a signature: `{ (): string }`

The function type form is generally preferred when possible for being more succinct.

This rule suggests using a function type instead of an interface or object type literal with a single call signature.

## Examples

<!--tabs-->

### âŒ Incorrect

```ts
interface Example {
>>>>>>> features/task-display-and-function
  (): string;
}
```

```ts
<<<<<<< HEAD
function foo(bar: { (): number }): number {
  return bar();
=======
function foo(example: { (): number }): number {
  return example();
>>>>>>> features/task-display-and-function
}
```

```ts
<<<<<<< HEAD
interface Foo extends Function {
  (): void;
}
```

```ts
interface MixinMethod {
=======
interface ReturnsSelf {
>>>>>>> features/task-display-and-function
  // returns the function itself, not the `this` argument.
  (arg: string): this;
}
```

<<<<<<< HEAD
Examples of **correct** code for this rule:

```ts
interface Foo {
  (): void;
  bar: number;
=======
### âœ… Correct

```ts
type Example = () => string;
```

```ts
function foo(example: () => number): number {
  return bar();
>>>>>>> features/task-display-and-function
}
```

```ts
<<<<<<< HEAD
=======
// returns the function itself, not the `this` argument.
type ReturnsSelf = (arg: string) => ReturnsSelf;
```

```ts
>>>>>>> features/task-display-and-function
function foo(bar: { (): string; baz: number }): string {
  return bar();
}
```

```ts
interface Foo {
  bar: string;
}
interface Bar extends Foo {
  (): void;
}
```

```ts
<<<<<<< HEAD
// returns the `this` argument of function, retaining it's type.
type MixinMethod = <TSelf>(this: TSelf, arg: string) => TSelf;
// a function that returns itself is much clearer in this form.
type ReturnsSelf = (arg: string) => ReturnsSelf;
```

```ts
=======
>>>>>>> features/task-display-and-function
// multiple call signatures (overloads) is allowed:
interface Overloaded {
  (data: string): number;
  (id: number): string;
}
// this is equivelent to Overloaded interface.
type Intersection = ((data: string) => number) & ((id: number) => string);
```

## When Not To Use It

If you specifically want to use an interface or type literal with a single call signature for stylistic reasons, you can disable this rule.

<<<<<<< HEAD
## Further Reading

- TSLint: [`callable-types`](https://palantir.github.io/tslint/rules/callable-types/)
=======
This rule has a known edge case of sometimes triggering on global augmentations such as `interface Function`.
These edge cases are rare and often symptomatic of odd code.
We recommend you use an [inline ESLint disable comment](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1).
See [#454](https://github.com/typescript-eslint/typescript-eslint/issues/454) for details.
>>>>>>> features/task-display-and-function
