# Suggest having hooks before any test cases (`prefer-hooks-on-top`)

<<<<<<< HEAD
All hooks should be defined before the start of the tests
=======
While hooks can be setup anywhere in a test file, they are always called in a
specific order which means it can be confusing if they're intermixed with test
cases.

This rule helps to ensure that hooks are always defined before test cases.
>>>>>>> features/task-display-and-function

## Rule Details

Examples of **incorrect** code for this rule

```js
/* eslint jest/prefer-hooks-on-top: "error" */

describe('foo', () => {
  beforeEach(() => {
<<<<<<< HEAD
    //some hook code
  });
  test('bar', () => {
    some_fn();
  });
  beforeAll(() => {
    //some hook code
  });
  test('bar', () => {
    some_fn();
  });
});

// Nested describe scenario
describe('foo', () => {
  beforeAll(() => {
    //some hook code
  });
  test('bar', () => {
    some_fn();
  });
  describe('inner_foo', () => {
    beforeEach(() => {
      //some hook code
    });
    test('inner bar', () => {
      some_fn();
    });
    test('inner bar', () => {
      some_fn();
    });
    beforeAll(() => {
      //some hook code
    });
    afterAll(() => {
      //some hook code
    });
    test('inner bar', () => {
      some_fn();
    });
  });
=======
    seedMyDatabase();
  });

  it('accepts this input', () => {
    // ...
  });

  beforeAll(() => {
    createMyDatabase();
  });

  it('returns that value', () => {
    // ...
  });

  describe('when the database has specific values', () => {
    const specificValue = '...';

    beforeEach(() => {
      seedMyDatabase(specificValue);
    });

    it('accepts that input', () => {
      // ...
    });

    it('throws an error', () => {
      // ...
    });

    afterEach(() => {
      clearLogger();
    });
    beforeEach(() => {
      mockLogger();
    });

    it('logs a message', () => {
      // ...
    });
  });

  afterAll(() => {
    removeMyDatabase();
  });
>>>>>>> features/task-display-and-function
});
```

Examples of **correct** code for this rule

```js
/* eslint jest/prefer-hooks-on-top: "error" */

describe('foo', () => {
<<<<<<< HEAD
  beforeEach(() => {
    //some hook code
  });

  // Not affected by rule
  someSetup();

  afterEach(() => {
    //some hook code
  });
  test('bar', () => {
    some_fn();
  });
});

// Nested describe scenario
describe('foo', () => {
  beforeEach(() => {
    //some hook code
  });
  test('bar', () => {
    some_fn();
  });
  describe('inner_foo', () => {
    beforeEach(() => {
      //some hook code
    });
    test('inner bar', () => {
      some_fn();
=======
  beforeAll(() => {
    createMyDatabase();
  });

  beforeEach(() => {
    seedMyDatabase();
  });

  afterAll(() => {
    clearMyDatabase();
  });

  it('accepts this input', () => {
    // ...
  });

  it('returns that value', () => {
    // ...
  });

  describe('when the database has specific values', () => {
    const specificValue = '...';

    beforeEach(() => {
      seedMyDatabase(specificValue);
    });

    beforeEach(() => {
      mockLogger();
    });

    afterEach(() => {
      clearLogger();
    });

    it('accepts that input', () => {
      // ...
    });

    it('throws an error', () => {
      // ...
    });

    it('logs a message', () => {
      // ...
>>>>>>> features/task-display-and-function
    });
  });
});
```
